In [9]:
   # repo URL : https://github.com/<redacted>/assignment-4-equation-of-a-slime-<redacted>
   # commit ID: 2610f92d7cb12365682817e0af63d12ab443b3ad

In [10]:
# Imports section
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import PolynomialFeatures
from sklearn.pipeline import Pipeline

## Part 1. Loading the dataset

In [11]:
# Using pandas load the dataset (load remotely, not locally)
df = pd.read_csv("/workspaces/assignment-4-equation-of-a-slime-<redacted>/science_data_large.csv")
# Output the first 15 rows of the data
print(df.head(15))
# Display a summary of the table information (number of datapoints, etc.)
print(df.info())

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

In [12]:
print(df.describe())

       Temperature °C     Mols KCL     Size nm^3
count     1000.000000  1000.000000  1.000000e+03
mean       500.500000   471.530000  5.086111e+05
std        288.819436   288.482872  4.474838e+05
min          1.000000     1.000000  1.611429e+01
25%        250.750000   226.750000  1.298267e+05
50%        500.500000   459.500000  3.827182e+05
75%        750.250000   710.250000  7.603211e+05
max       1000.000000  1000.000000  1.972127e+06


## Part 2. Splitting the dataset

In [13]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = df.drop("Size nm^3", axis=1)
y = df["Size nm^3"]
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size = 0.9, test_size=0.1, random_state=42)

## Part 3. Perform a Linear Regression

In [14]:
# Use sklearn to train a model on the training set
model = LinearRegression()
model.fit(X_train, y_train)
# Create a sample datapoint and predict the output of that sample with the trained model
# sample_dataPoint = [[95,499]]
sample_dataPoint = pd.DataFrame([[95, 499]], columns=["Temperature °C", "Mols KCL"])
predicted_dataPoint = model.predict(sample_dataPoint)
print("predicted_dataPoint: ", predicted_dataPoint)
# Report on the score for that model, in your own words (markdown, not code) explain what the score means
model_score = model.score(X_test, y_test)
print("model_score: ", model_score)
# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
coefficients = model.coef_
intercept = model.intercept_

print("Coefficients:", coefficients)
print("Intercept:", intercept)

predicted_dataPoint:  [188207.26786334]
model_score:  0.8552472077276095
Coefficients: [ 866.14641337 1032.69506649]
Intercept: -409391.47958340764


Sample equation:  $E = MC^2$ 

The R² score ranges from 0 to 1:
- A score of 1 indicates that the model explains all the variance in the target variable, meaning it is a perfect fit.

The model's R² score is 0.855, which means it explains 85.5% of the variation in the target variable. This is a good score, showing that the model does a good job predicting "Size $nm^3$" based on the features (Temperature and Mols KCL). However, it's not perfect, so there may still be some patterns or information that the model isn't capturing completely.

The model equtation is given by $h(x) = \theta_0 + \theta_1 \cdot x_1 + \theta_2 \cdot x_2$

$h(x) = -409391.48 + 866.15 \cdot x_1 + 1032.70 \cdot x_2$ 

OR

$h(x) = -409391.48 + 866.15 \cdot  Temperature °C  + 1032.70 \cdot Mols KCL$


## Part 4. Use Cross Validation

In [15]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data

# Performing cross-validation with 5 folds.
scores = cross_val_score(model, X, y, cv=5)

# Output the cross-validation scores and the mean score
print("Cross-validation scores:", scores)
print("Mean of cross-validation score:", scores.mean())

# Report on their finding and their significance
# Report given in the Markdown.

Cross-validation scores: [0.83918826 0.87051239 0.85871066 0.87202623 0.84364641]
Mean of cross-validation score: 0.8568167899144437


The individual scores show how well the model worked on different parts of the data after it was shuffled. Since the scores are close to 1, it means the model does a good job of explaining most of the differences in the data across these different shuffles, which suggests it works well for different sets of data.


The average score is 0.857, meaning that on average, the model explains about 85.7% of the variations in the target variable. This is a strong result for a regression model and indicates that the model has learned important patterns from the data.

## Part 5. Using Polynomial Regression

In [16]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Step 1: Creating a pipeline with PolynomialFeatures and LinearRegression
pipeline = Pipeline([
    ('poly', PolynomialFeatures(degree=2)),
    ('linear', LinearRegression())
])

pipeline.fit(X_train, y_train)
# printing the R-squared score
r2_score = pipeline.score(X_test, y_test)
print("R-squared score:", r2_score)
sample_dataPoint = pd.DataFrame([[95, 499]], columns=["Temperature °C", "Mols KCL"])
predicted_dataPoint = pipeline.predict(sample_dataPoint)
print("predicted_dataPoint: ",predicted_dataPoint)
# Report on the metrics and output the resultant equation as you did in Part 3.
coefficients = pipeline.named_steps['linear'].coef_
intercept = pipeline.named_steps['linear'].intercept_

# # printing the coefficients and intercept
print("Coefficients:", coefficients)
print("Intercept:", intercept)
# Getting the output feature names for transformation.
feature_names = pipeline.named_steps['poly'].get_feature_names_out()
print("Feature Names:", feature_names)

# wiriting the equation of h(x) in the markdown.

R-squared score: 1.0
predicted_dataPoint:  [103064.3142757]
Coefficients: [ 0.00000000e+00  1.20000000e+01 -1.27195488e-07  1.26494371e-11
  2.00000000e+00  2.85714287e-02]
Intercept: 2.0477105863392353e-05
Feature Names: ['1' 'Temperature °C' 'Mols KCL' 'Temperature °C^2'
 'Temperature °C Mols KCL' 'Mols KCL^2']


The R² score is 1.0, meaning the model perfectly explains all the variance in the target variable (Size nm³) based on the features. A score of 1 indicates a perfect fit, meaning the model captures the relationship between the features and the target variable without any unexplained variation.

I noticed that the "Size $nm^3$" predicted by the linear regression model was slightly different from the value given in the dataset for the same data point. However, when I used the same data point with the polynomial regression model, it predicted the exact same "Size $nm^3$" value as in the dataset.

This suggests that the polynomial regression model is better at capturing the underlying relationship between the features (Temperature °C and Mols KCL) and the target variable ("Size $nm^3$")


$h(x) = 2.0477 + \theta_1 \cdot x_1 + \theta_2 \cdot x_2 + \theta_3 \cdot x_1^2 + \theta_4 \cdot x_1 \cdot x_2 + \theta_5 \cdot x_2^2$

$h(x) = 2.0477 \cdot 10^{-5} + 12 \cdot Temperature °C - 1.27 \cdot 10^{-7} \cdot Mols KCL + 1.26 * 10^{-11} \cdot Temperature °C^2 + 2 \cdot Temperature °C \cdot Mols KCL + 2.857 * 10^{-2} \cdot Mols KCL^2$

Since we can ignore values outside of 5 significant figures (e.g. 0.000003 is just 0) our $h(x)$ is

$h(x) = 2.0477\cdot 10^{-5} + 12 \cdot Temperature °C + 2 \cdot Temperature °C \cdot Mols KCL + 2.857 * 10^{-2} \cdot Mols KCL^2$

The equation shows a combination of linear, squared, and interaction terms between the features, capturing both the individual and joint effects of Temperature °C and Mols KCL on the target variable.